src/main/scala/impl/group5/NumberOfLoopsAnalysis.scala

```diff
@@ -0,0 +1,128 @@
+package org.tud.sse.metrics
+package impl.group5
+
+import analysis.{ClassFileAnalysis, MetricValue}
+import input.CliParser.OptionMap
+
+import org.opalj.br.ClassFile
+import org.opalj.br.analyses.Project
+
+import java.net.URL
+import scala.collection.mutable.ListBuffer
+import scala.util.Try
+
+/**
+ * Die Klasse NumberOfLoopsAnalysis stellt die Implementierung  der Metrik Number of Loops (Loop) da.
+ * Diese Metrik zählt für jede Methode die Anzahl der Schleifen und gibt sie aus.
+ *
+ * Die Default Einstellung ist, dass nur die Methoden ausgeben werden, in denen sich Loops befinden. Die anderen Methoden in denen Loops 0 ist werden weggelassen
+ *
+ * Optional CLI argument:
+ *  - --only-class Es wird nur die Anzahl an Loops in einer Klasse angezeigt und nicht für jede Methode
+ *  - --out-all-loops Es werden alle Methoden ausgeben, auch die mit Loop 0.
+ *  - --no-class Es wird nicht die gesamte Zahl der Loops einer Klasse ausgeben
+ *  - --only-class-with-loops Gibt nur Klassen mit Loops aus.
+ */
+
+class NumberOfLoopsAnalysis extends ClassFileAnalysis {
+
+  private val onlyClassSymbol: Symbol = Symbol("only-class")
+  private val outAllSymbol: Symbol = Symbol("out-all-loops")
+  private val noClassSymbol: Symbol = Symbol("no-class")
+  private  val onlyClassWithLoopsSymbol: Symbol = Symbol("only-class-with-loops")
+
+  /**
+   * Die Methode führt die Metrik Loop aus.
+   * @param classFile Ist ein Interface von Projekt
+   * @param project Bekommt die vom Framework gelesende jar file in einer Präsentationsform
+   * @param customOptions Einstellungs Möglichkeiten der Analyse
+   * @return Das Ergebniss wird in der Liste für die Ausgabe gespeichert
+   */
+  override def analyzeClassFile(classFile: ClassFile, project: Project[URL], customOptions: OptionMap): Try[Iterable[MetricValue]] = Try {
+
+    val outAllMethode = customOptions.contains(outAllSymbol)
+    val onlyClass = customOptions.contains(onlyClassSymbol)
+    val noClass = customOptions.contains(noClassSymbol)
+    val onClassLoops = customOptions.contains(onlyClassWithLoopsSymbol)
+
+    if(noClass &&(onlyClass||onClassLoops))
+      {
+        log.error("Fehler! Argumente sind inkompatibel")
+      }
+
+    val methods= classFile.methodsWithBody
+    var classloops = 0
+    val rlist = new ListBuffer[MetricValue]()
+
+    while(methods.hasNext) {
+      val method = methods.next()
+      val body = method.body
+      var loops = 0
+      val line = new ListBuffer[Int]()
+      if (body.nonEmpty) {
+        val code = body.get
+        val instructions = code.instructions
+        var linelegth = 0
+        for (instruction <- instructions) {
+          // Im Array vom Framework können leere Felder sein, die mit null belegt sind
+          if (instruction != null) {
+            //opcode von goto ist 167
+            if (instruction.opcode == 167) {
```

Comment:
Opcode 167 ist nicht der einzige Opcode, der für die Herstellung von Schleifen im Bytcode von den Kompilern verwendet wird. Insonderen werden auch die Instruktionen "if_icmp<cond>" (https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-6.html) verwandt. Beispiele für solche Schleifenkonstruktionen sind hier zu finden:  https://blog.jamesdbloom.com/JavaCodeToByteCode_PartOne.html#loops , https://stackoverflow.com/questions/6792305/identify-loops-in-java-byte-code

Reply:
:smiley::thumbsup: